### PR TITLE
Release tags store entries when a check sampler is discarded

### PR DESCRIPTION
**Pull request by vickenty**
### What does this PR do?

Release tags store entries when a check sampler is discarded.

When a check is removed, contexts for active metrics are not expired yet, and are not removed from the context resolver, but the entire object tree is simply discarded. This patch adds a finalizer to context resolver, so reference counts can be updated for context objects that are about to be GC'd.

### Motivation

Avoid accumulating tags from checks that were de-scheduled.

### Possible Drawbacks / Trade-offs

Finalizers are reliable for the purpose, but timing is unpredictable. Explicitly clearing the resolver on check deregister would be more deterministic, but requires plumbing a new call through several components.

### Describe how to test/QA your changes

Run the agent with tags store enabled, and a large number of checks scheduled, sending metrics, and de-scheduled.

Verify that total number of entries in the tags store does grow and remains similar to the number of metrics sent.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/aggregator/tags/store.go [hunk 1]
```diff
@@ -30,8 +31,11 @@ func (e *Entry) Tags() []string {
 
 // Release decrements internal reference counter, potentially marking
 // the entry as unused.
+//
+// Can be called concurrently with Insert and Shrink.
 func (e *Entry) Release() {
-	e.refs--
+	const minusOne = ^uint64(0)
+	atomic.AddUint64(&e.refs, minusOne)
```
Comment (djmitche) [inline pkg/aggregator/tags/store.go:38]: I suspect that switching to atomic operations here is going to cause a lot more bus traffic and make this quite a bit slower.  How has this fared on the benchmark env?

**Review by djmitche — COMMENTED** [review]
I'm pretty wary of starting to use finalizers like this.  From my (limited) understanding of Go finalizers, it seems like this would work, so I'm not saying it's incorrect.  But I'm worried about setting a precedent of using a fairly dangerous tool, and I'm worried about debugging this if it _doesn't_ work.

To date, it seems we only use finalizers for FFI and the like (python checks and EBPF).  Everywhere else, we use explicit Close(), Stop(),  `defer` and the like.  How difficult would it be to make this situation explicit, too?

File: pkg/aggregator/tags/store.go (hunk 1 above)
Comment (vickenty) [reply pkg/aggregator/tags/store.go:38]: No change, this doesn't happen frequently enough to make a noticeable difference. Even on a metric churn test this doesn't show up in the cpu profiles at all.

Comment (djmitche) [reply pkg/aggregator/tags/store.go:38]: I think you'd see it in overall CPU usage - the atomic operation will clear caches in all CPUs, and cause increased bus contention, so you'd see a general slowdown.  But if you've not seen that sort of result, great!

**Comment by djmitche** [comment]
That alternative doesn't seem bad to me, but I suppose it's a matter of taste.  This is also one of the few places we're using atomic refcounts, so perhaps it makes sense to introduce this as a new idiom in the codebase.

File: pkg/aggregator/tags/store.go (hunk 1 above)
Comment (vickenty) [reply pkg/aggregator/tags/store.go:38]: My understanding is that atomics do have such drastic effects, at least on modern x64, and in most cases can be performed without any penalty at all.

**Pull request closed without merging**